### PR TITLE
Feature/84 add food collection page read function 1

### PR DESCRIPTION
**Pull request by kotaaaa**
## Overview

- Added food collection page 

## Functions
- Users can filter by group, category, container name and any customized query at filter modal.
- Users can select category that users want to filter by from category selection modal 
- Users can sort by name, createdAt, expire date.

## Outside the scope of this PR
- Did not apply selection modal component (#114 ) yet 
- Attach small red batch to filter icon when any filter is applied
- Drawer title of "Category" at FoodFilterCategoryDrawer.tsx is not `text-center`
- `All clear` function at FoodFilterCategoryDrawer.tsx
- Fix the position of `SearchBar`

## Screen Captures

https://github.com/genesis-tech-tribe/nishiki-frontend/assets/25422441/1c81725c-8f1d-4a61-89d9-1ccd32c64afd

## Reviewer Checklist:

- [ ] If css from Figma is applied properly or not.
- [ ] If overall html structure is organized or not.
- [ ] If code is organized well and there is any room to improve in the point of readability and performance or not.

### CODE REVIEW

File: src/components/page/FoodsPage.tsx [hunk 1]
```diff
@@ -1,3 +1,156 @@
...
+      <div className="relative">
+        <SearchBar />
+        <FilterButton
+          containers={containersGroupByGroups}
+          containerIdMap={containerIdMap}
+          groupIdMap={groupIdMap}
+        />
+      </div>
+      <div
+        className={`flex ${
```
Comment (kanta1207) [inline src/components/page/FoodsPage.tsx:112]: ### (Optional)
The code here is having no problem, my suggestion here is just optional !

I often like using twMerge(in our project, we can utilize `cn()` method from `@/lib/tailwind/utils`) to avoid using template literals inside of className. 
Because, when you're using template literal inside of className, just missing a single space can easily cause an unexpected behaviour.

If you use `cn()` method in here, it would be like this ↓
`   className={cn(
          'flex',
          group ||
            container ||
            (categoryList.length > 0 &&
              'gap-1.5 mt-4 mb-1.5 sm:overflow-auto sm:whitespace-normal overflow-scroll whitespace-nowrap'),
        )}`
      
This is not the main purpose of using `twMerge` tho, plz take a look and consider it 🙏

File: src/components/page/FoodsPage.tsx [hunk 2]
```diff
@@ -1,3 +1,156 @@
...
+  groupContainersByGroupId,
+} from '@/features/foods/utils/containerMapping';
+import { IContainer, IFood } from '@/types/definition';
+
+import { Route } from 'next';
+import { usePathname, useRouter, useSearchParams } from 'next/navigation';
+import React, { useEffect, useState } from 'react';
+
+export const FoodsPage = ({ containers }: { containers: IContainer[] }) => {
+  const pathname = usePathname();
```
Comment (kanta1207) [inline src/components/page/FoodsPage.tsx:24]: ### (Optional)
Great job on implementing all the filtering logic!

Creating one or two custom hooks to separate the logic out of this file would make it even better tho, 
are you gonna do it in different scope?

File: src/features/foods/components/FoodSort.tsx [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        </SelectionDrawerTrigger>
+        <SelectionDrawerContent>
+          <SelectionDrawerRadioGroup defaultValue={selectedOption} onValueChange={onValueChange}>
+            {sortOptions.map((option, i) => {
+              return (
+                <Label
+                  key={i}
+                  htmlFor={option.id}
+                  variant="selectionDrawer"
+                  className={cn(i !== 0 && 'border-t border-gray-light')}
```
Comment (kanta1207) [inline src/features/foods/components/FoodSort.tsx:73]: Oh now I found in here you're actually utilizing` cn()` method haha. I like this

**Review by kanta1207 — COMMENTED** [review]
Thank you for implementing all the complex filtering logic!

My suggestions/questions are all optional ones. Appreciated if I can hear your point of view on them.

I'll not approve it yet to wait for @uskayyyyy to take a look at them, but for me couldn't find any big problems!

File: src/components/page/FoodsPage.tsx [hunk 4]
```diff
@@ -1,3 +1,156 @@
...
+              onCrossClick={() => removeCategoryFilter(key)}
+            />
+          );
+        })}
+      </div>
+      <div className="flex items-center justify-end">
+        <FoodSort />
+        <div className="px-3.5">
+          <Icon icon={MenuMeatballIcon} size={4} />
+        </div>
```
Comment (uskayyyyy) [inline src/components/page/FoodsPage.tsx:151]: This meatball button should take up 48x48 px:

image

image

File: src/features/foods/components/FoodSort.tsx [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+      params.delete('sort');
+    }
+    replace(`${pathname}?${params.toString()}` as Route);
+  };
+
+  return (
+    <>
+      <SelectionDrawerRoot open={isDrawerOpen} onOpenChange={setIsDrawerOpen}>
+        <SelectionDrawerTrigger asChild>
+          <button className="flex items-center gap-2">
```
Comment (uskayyyyy) [inline src/features/foods/components/FoodSort.tsx:60]: There is an unwanted gap, which does not exist in Figma:
image

**Review by uskayyyyy — CHANGES_REQUESTED** [review]
@kotaaaa I have reviewed the UI part and left some comments. Please have a look at them.

I will proceed with reviewing the filter part after this.

File: src/features/foods/components/FoodSort.tsx (hunk 5 above)
Comment (uskayyyyy) [reply src/features/foods/components/FoodSort.tsx:60]: This text must match the selected sort.

image

File: src/components/page/FoodsPage.tsx [hunk 6]
```diff
@@ -1,3 +1,156 @@
...
+        <FilterButton
+          containers={containersGroupByGroups}
+          containerIdMap={containerIdMap}
+          groupIdMap={groupIdMap}
+        />
+      </div>
+      <div
+        className={`flex ${
+          group || container || categoryList.length > 0
+            ? 'gap-1.5 mt-4 mb-1.5 sm:overflow-auto sm:whitespace-normal overflow-scroll whitespace-nowrap'
```
[the remark this answers is not in the document]
Comment (uskayyyyy) [reply src/components/page/FoodsPage.tsx:114]: Could you also replace the margin-bottom with padding-bottom? 

image

Because the scroll bar overlays the badge.
Screen Recording 2023-12-21 at 19.23.08.mov

File: src/features/foods/components/FilterBadge.tsx [hunk 7]
```diff
@@ -29,7 +29,7 @@ const FilterBadge = ({
           <CircledIcon />
         </div>
       ) : emoji ? (
-        <div className="bg-white w-4 h-4 rounded-full p-[3px] mr-1 flex items-center justify-center text-2xs select-none">
+        <div className="bg-white w-4 h-4 rounded-full p-[3px] mr-1 flex items-center justify-center text-xs select-none">
```
Comment (uskayyyyy) [inline src/features/foods/components/FilterBadge.tsx:32]: Why is this change needed? It looks bigger than the Figma design now...

File: src/components/page/FoodsPage.tsx (hunk 1 above)
Comment (uskayyyyy) [reply src/components/page/FoodsPage.tsx:112]: After experimenting, I figured out that giving the following styles works perfectly
```
    margin-left: -16px;
    margin-right: -16px;
    padding-left: 16px;
    padding-right: 16px;
    padding-bottom: 4px;
```

**Review by uskayyyyy — CHANGES_REQUESTED** [review]
@kotaaaa 
I have reviewed the filter part and was impressed by how you implemented the filtering logic! I left some comments, so please kindly take a look at them.

However, I felt it needed to be refactored, especially taking out some code as custom hooks. If you don't mind, I would like to try it.

File: src/utils/dateUtils.ts
[the remark this answers is not in the document]
Comment (kotaaaa) [reply src/utils/dateUtils.ts:1]: > src/features/foods/types/utils.ts

I changed this name to `src/features/foods/types/FoodViewModel.ts`.
It seemed too much `utils.ts` file as you mentioned.

File: src/types/definition.ts
[the remark this answers is not in the document]
Comment (kotaaaa) [reply src/types/definition.ts:1]: As I discussed with backend side [here](https://github.com/genesis-tech-tribe/nishiki-documents/issues/89), 
`/container` REST response object includes `groupId` and `groupName`.
Currently container is used with group information. 
So (It might a little strange though), I thought we should create another pure `container` interface from current `IContainer` when we need. 

In addition, I proposed to change container REST response schema [here](https://github.com/genesis-tech-tribe/nishiki-documents/issues/111), so if this is accepted, we can include `IGroup` object to `IContainer` so that it will be a little more organized.

File: src/components/page/FoodsPage.tsx [hunk 8]
```diff
@@ -1,3 +1,156 @@
...
+  const sort = searchParams?.get('sort') || '';
+  const group = searchParams?.get('group') || '';
+  const container = searchParams?.get('container') || '';
+  const [displayedFoods, setDisplayedFoods] = useState<IFoodView[]>([]);
+  const [query, setQuery] = useState<string>(searchParams?.get('query') || '');
+  const [categoryList, setCategoryList] = useState<string[]>(
+    searchParams?.get('category')?.split(',') || [],
+  );
+
+  const containersGroupByGroups = groupContainersByGroupId(containers);
```
[the remark this answers is not in the document]
Comment (kotaaaa) [reply src/components/page/FoodsPage.tsx:37]: Thank you for your suggestion!!
I agree with your opinion to be more understandable by setting `GroupId` instead of `Groups`.
So I would like to choose this.

- `containersGroupByGroupId`

> It should be at least "grouped", not "group"

In SQL grammar, we use `GROUP BY` so I thought we can easily grasp this meaning with it. and it causes to make the variable longer. How do you think?
https://www.w3schools.com/sql/sql_groupby.asp#:~:text=The%20SQL%20GROUP%20BY%20Statement,by%20one%20or%20more%20columns.

File: src/components/page/FoodsPage.tsx [hunk 9]
```diff
@@ -1,3 +1,156 @@
...
+  const containerIdMap = createContainerIdNameMap(containers);
+  const groupIdMap = createGroupIdNameMap(containers);
+  const updateUrlParams = (key: string, value?: string | string[]) => {
```
[the remark this answers is not in the document]
Comment (kotaaaa) [reply src/components/page/FoodsPage.tsx:40]: Thank you!! Fixed. Do you know we can add this rule to `tslint` or `prettier`? (We might be able to put this to github issues.)

File: src/features/foods/utils/containerMapping.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+export const groupContainersByGroupId = (data: IContainer[]): Record<string, string[]> => {
+  return data.reduce(
+    (acc, { groupId, id }) => ({
+      ...acc,
+      [groupId]: acc[groupId] ? [...acc[groupId], id] : [id],
+    }),
+    {} as Record<string, string[]>,
+  );
+};
```
[the remark this answers is not in the document]
Comment (kotaaaa) [reply src/features/foods/utils/containerMapping.ts:11]: Thank you for your notice, I created common type and put them to `FoodTypes.ts`.

File: src/features/foods/utils/containerMapping.ts
[the remark this answers is not in the document]
Comment (kotaaaa) [reply src/features/foods/utils/containerMapping.ts:1]: Thank you!! I changed to `containers`.
I thought the word of `container` were a lot in this file. I changed some other word(container -> c) to to keep the readability.

File: src/utils/dateUtils.ts
[the remark this answers is not in the document]
Comment (kotaaaa) [reply src/utils/dateUtils.ts:1]: > In my working branch for the Calendar component, I use an npm package called [date-fns](https://www.npmjs.com/package/date-fns) to format a Date object. The following code that specifies "PP" converts a Date into this format:Apr 29, 1453`. Which do you think is better, using the package or having a custom function?


Thank you for your suggestion! for date matters, I think utilizing library is better and hassle-free. I changed to use them.

File: src/features/foods/components/FoodList.tsx
[the remark this answers is not in the document]
Comment (kotaaaa) [reply src/features/foods/components/FoodList.tsx:1]: I created issue https://github.com/genesis-tech-tribe/nishiki-frontend/issues/118, I will handle there.

**Comment by kotaaaa** [comment]
> The badges' text breaks unwantedly when the screen width increases. Could you investigate whether the cause is ascribed to this PR. We might need a separate issue to handle this though.
I created issue https://github.com/genesis-tech-tribe/nishiki-frontend/issues/117, I will investigate this with this ticket

File: src/components/page/FoodsPage.tsx (hunk 1 above)
Comment (kotaaaa) [reply src/components/page/FoodsPage.tsx]: > The badges' text breaks unwantedly when the screen width increases. Could you investigate whether the cause is ascribed to this PR. We might need a separate issue to handle this though.

I created issue https://github.com/genesis-tech-tribe/nishiki-frontend/issues/117, I will investigate this with this ticket

**Comment by kotaaaa** [comment]
@kanta1207  @uskayyyyy  
Thank you guys for you kind reviews!!
I checked all of them and fixed or commented or created new issues to handle them!!
Please kindly check my revisions and if I miss anything or additional points, please let me know!

File: src/features/foods/types/FoodTypes.ts [hunk 11]
```diff
@@ -0,0 +1,7 @@
+import { IContainer, IFood } from '@/types/definition';
+
+export interface IFoodView extends IFood {
+  container: string;
+}
+export type GroupIdContainersMapType = Record<IContainer['groupId'], IContainer['id'][]>;
```
Comment (kotaaaa) [inline src/features/foods/types/FoodTypes.ts:6]: I thought `GroupIdContainersMapType` should be in `FoodTypes.ts` as this is referred from more than one file (`FilterButton.tsx` and `containerMappings.tsx`)

File: src/features/foods/components/BadgeList.tsx [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    setCategoryList(updatedCategoryList);
+    updateUrlParams('category', updatedCategoryList);
+  };
+
+  return (
+    <div
+      className={cn(
+        'flex',
+        (group || container || categoryList.length > 0) && 'gap-1.5 mt-4 pb-1.5',
+        'overflow-x-auto margin-right: -16px; margin-right: -16px; padding-left: 16px; padding-right: 16px; padding-bottom: 4px;',
```
Comment (uskayyyyy) [inline src/features/foods/components/BadgeList.tsx:51]: These should never work 😂 Please replace them with tailwind classes.

File: src/components/page/FoodsPage.tsx (hunk 2 above)
Comment (uskayyyyy) [reply src/components/page/FoodsPage.tsx]: Moved to #121 [](url)

File: src/components/page/FoodsPage.tsx (hunk 1 above)
Comment (kotaaaa) [reply src/components/page/FoodsPage.tsx]: Finally, I set these css for this. Could you please check at your enviroment? 
I thought #117 is also solved. 

> flex -mx-4 px-4 gap-1.5 mt-4 pb-1 overflow-x-auto whitespace-nowrap

File: src/components/page/FoodsPage.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (uskayyyyy) [reply src/components/page/FoodsPage.tsx]: I researched and found a configuration to add a line break before every function definition. But, it works only with the `function` keyword, not with arrow functions.

```json
// .eslintrc
{
  "extends": ["prettier"],
  "plugins": ["prettier"],
  "rules": {
    "prettier/prettier": "error",
    "padding-line-between-statements": [
      "error",
      { "blankLine": "always", "prev": "*", "next": "function" }
    ]
  }
}
```

[the remark this answers is not in the document]
Comment (kotaaaa) [reply src/components/page/FoodsPage.tsx]: Woo, thank you for your fast work!! got it!!

File: src/features/foods/utils/containerMapping.ts
[the remark this answers is not in the document]
Comment (uskayyyyy) [reply src/features/foods/utils/containerMapping.ts:1]: I like that way!

File: src/components/page/FoodsPage.tsx (hunk 1 above)
Comment (uskayyyyy) [reply src/components/page/FoodsPage.tsx]: I appreciate it!

**Review by uskayyyyy — APPROVED** [review]
@kotaaaa 
Thank you for fixing everything! It now looks perfect😆

File: src/features/foods/components/FoodList.tsx
[the remark this answers is not in the document]
Comment (uskayyyyy) [reply src/features/foods/components/FoodList.tsx:1]: I deleted the issue because it's solved in this PR

**Comment by kotaaaa** [comment]
Thank you guys!!
